### PR TITLE
typescript type definition

### PR DESCRIPTION
**Pull request by shumih**
Types based on current api

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,1353 @@
...
+     * Device's custom manufacturer data. Its format is defined by manufacturer.
+     * @private
+     */
+    manufacturerData?: Base64;
+
+    /**
+     * Map od service UUIDs with associated data.
+     * @private
+     */
+    serviceData?: { [uuid: UUID]: Base64 };
```
Comment (Cierpliwy) [inline index.d.ts]: `serviceData: { [uuid: string]: Base64 } | null;`

Comment (Cierpliwy) [reply index.d.ts]: We need to use `string` here directly because type aliases are not accepted as indices.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,1353 @@
...
+
+    /**
+     * List of available services visible during scanning.
+     * @private
+     */
+    serviceUUIDs?: UUID[];
```
Comment (Cierpliwy) [inline index.d.ts]: `serviceUUIDs: UUID[] | null`

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,1353 @@
...
+
+    /**
+     * User friendly name of device.
+     * @private
+     */
+    localName?: string;
```
Comment (Cierpliwy) [inline index.d.ts]: `localName: string | null`

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,1353 @@
...
+
+    /**
+     * Transmission power level of device.
+     * @private
+     */
+    txPowerLevel?: number;
```
Comment (Cierpliwy) [inline index.d.ts]: `txPowerLevel: number | null`

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,1353 @@
...
+
+    /**
+     * List of solicited service UUIDs.
+     * @private
+     */
+    solicitedServiceUUIDs?: UUID[];
```
Comment (Cierpliwy) [inline index.d.ts]: `solicitedServiceUUIDs: UUID[] | null`

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,1353 @@
...
+
+    /**
+     * Is device connectable.
+     * @private
+     */
+    isConnectable?: boolean;
```
Comment (Cierpliwy) [inline index.d.ts]: `isConnectable: boolean | null`

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,1353 @@
...
+
+    /**
+     * List of overflow service UUIDs.
+     * @private
+     */
+    overflowServiceUUIDs?: UUID[];
```
Comment (Cierpliwy) [inline index.d.ts]: `overflowServiceUUIDs: UUID[] | null`

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,1353 @@
...
+    /**
+     * True if characteristic is monitoring value changes with ACK.
+     * @private
+     */
+    isIndicatable: boolean;
+    /**
+     * Characteristic value if present
+     * @private
+     */
+    value?: Base64;
```
Comment (Cierpliwy) [inline index.d.ts]: `value: Base64 | null;`

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,1353 @@
...
+    monitorCharacteristicForDevice(
+      deviceIdentifier: Device['id'],
+      serviceUUID: UUID,
+      characteristicUUID: UUID,
+      listener: (error?: BleError, characteristic?: Characteristic) => void,
+      transactionId?: string
+    ): Subscription;
+  }
+
+  export const EventEmitter = NativeEventEmitter;
```
Comment (Cierpliwy) [inline index.d.ts]: Not needed.

File: index.d.ts [hunk 10]
```diff
@@ -0,0 +1,1353 @@
...
+     */
+    serviceID: Identifier;
+    /**
+     * Service's UUID to which characteristic belongs
+     */
+    serviceUUID: UUID;
+    /**
+     * Device's ID to which characteristic belongs
+     */
+    deviceID: DeviceId;
```
Comment (Cierpliwy) [inline index.d.ts]: Not declared. Currently we have mixed usage of `Device['id']` and `string`. Maybe it's better to go back to `DeviceId` alias and declare it?

File: index.d.ts [hunk 11]
```diff
@@ -0,0 +1,1353 @@
...
+     */
+    isNotifying: boolean;
+    /**
+     * True if characteristic is monitoring value changes with ACK.
+     */
+    isIndicatable: boolean;
+    /**
+     * Characteristic value if present
+     */
+    value?: Base64;
```
Comment (Cierpliwy) [inline index.d.ts]: `value: Base64 | null;`

File: index.d.ts [hunk 12]
```diff
@@ -0,0 +1,1353 @@
...
+    androidErrorCode?: BleAndroidErrorCode;
+    /**
+     * Platform specific error message.
+     */
+    reason?: string;
+
+    constructor(nativeBleError: NativeBleError | string, errorMessageMapping: BleErrorCodeMessageMapping);
+  }
+
+  export class Device {
```
Comment (Cierpliwy) [inline index.d.ts]: Please declare in the same way as `class Service`:
* `implements NativeDevice`
* Add methods
* All properties are always present (remove `?`) and add `null` to the type if `?` was present.

File: index.d.ts [hunk 13]
```diff
@@ -0,0 +1,1353 @@
...
+
+    /**
+     * Current, global {@link State} of a {@link BleManager}. All APIs are working only when active state
+     * is "PoweredOn".
+     *
+     * @returns {Promise<State>} Promise which emits current state of BleManager.
+     */
+    state(): Promise<State> {
+      return this._callPromise(BleModule.state());
+    }
```
Comment (Cierpliwy) [inline index.d.ts]: Remove implementation.

File: index.d.ts [hunk 14]
```diff
@@ -0,0 +1,1353 @@
...
+     *      }
+     *  }, true);
+     *
+     * @param {function(newState: State)} listener Callback which emits state changes of BLE Manager.
+     * Look at {@link State} for possible values.
+     * @param {boolean} [emitCurrentState=false] If true, current state will be emitted as well. Defaults to false.
+     *
+     * @returns {Subscription} Subscription on which `remove()` function can be called to unsubscribe.
+     */
+    onStateChange(listener: (newState: State) => void, emitCurrentState: boolean = false): Subscription;
```
Comment (Cierpliwy) [inline index.d.ts]: `onStateChange(listener: (newState: State) => void, emitCurrentState?: boolean): Subscription;`
Initializer is not allowed.

File: index.d.ts [hunk 15]
```diff
@@ -0,0 +1,1353 @@
...
+     * scanned {@link Device}. If `null` is passed, all available {@link Device}s will be scanned.
+     * @param {?ScanOptions} options Optional configuration for scanning operation.
+     * @param {function(error?: BleError, scannedDevice: ?Device)} listener Function which will be called for every scanned
+     * {@link Device} (devices may be scanned multiple times). It's first argument is potential {@link Error} which is set
+     * to non `null` value when scanning failed. You have to start scanning process again if that happens. Second argument
+     * is a scanned {@link Device}.
+     */
+    startDeviceScan(
+      UUIDs: UUID[] | null,
+      options: Partial<ScanOptions>,
```
Comment (Cierpliwy) [inline index.d.ts]: Partial is not needed after fixing ScanOptions: `options: ScanOptions | 'null',`

File: index.d.ts [hunk 16]
```diff
@@ -0,0 +1,1353 @@
...
+      listener: (error?: BleError, scannedDevice?: Device) => void
```
Comment (Cierpliwy) [inline index.d.ts]: `listener: (error: BleError | null, scannedDevice: Device | null) => void`

File: index.d.ts [hunk 17]
```diff
@@ -0,0 +1,1353 @@
...
+     *
+     * @param {Device['id']} deviceIdentifier {@link Device} identifier to be monitored.
+     * @param {function(error?: BleError, device: Device)} listener - callback returning error as a reason of disconnection
+     * if available and {@link Device} object. If an error is null, that means the connection was terminated by
+     * {@link #blemanagercanceldeviceconnection|bleManager.cancelDeviceConnection()} call.
+     * @returns {Subscription} Subscription on which `remove()` function can be called to unsubscribe.
+     */
+    onDeviceDisconnected(
+      deviceIdentifier: Device['id'],
+      listener: (error?: BleError, device: Device) => void
```
Comment (Cierpliwy) [inline index.d.ts]: `listener: (error: BleError | null, device: Device) => void`

Comment (Cierpliwy) [reply index.d.ts]: Fix docs as well.

File: index.d.ts [hunk 18]
```diff
@@ -0,0 +1,1353 @@
...
+     * {@link Characteristic} objects with modified value for each notification.
+     * @param {?string} transactionId optional `transactionId` which can be used in
+     * {@link #blemanagercanceltransaction|cancelTransaction()} function.
+     * @returns {Subscription} Subscription on which `remove()` function can be called to unsubscribe.
+     */
+    monitorCharacteristicForDevice(
+      deviceIdentifier: Device['id'],
+      serviceUUID: UUID,
+      characteristicUUID: UUID,
+      listener: (error?: BleError, characteristic?: Characteristic) => void,
```
Comment (Cierpliwy) [inline index.d.ts]: `listener: (error: BleError | null, characteristic: Characteristic | null) => void`

File: index.d.ts [hunk 19]
```diff
@@ -0,0 +1,1353 @@
...
+     * @param {UUID} characteristicUUID - {@link Characteristic} UUID.
+     * @param {function(error?: BleError, characteristic?: Characteristic)} listener callback which emits
+     * {@link Characteristic} objects with modified value for each notification.
+     * @param {?TransactionId} transactionId optional `transactionId` which can be used in
+     * {@link #blemanagercanceltransaction|bleManager.cancelTransaction()} function.
+     * @returns {Subscription} Subscription on which `remove()` function can be called to unsubscribe.
+     */
+    monitorCharacteristic(
+      characteristicUUID: UUID,
+      listener: (error?: BleError, characteristic?: Characteristic) => void,
```
Comment (Cierpliwy) [inline index.d.ts]: `listener: (error: BleError | null, characteristic: Characteristic | null) => void`

File: index.d.ts [hunk 20]
```diff
@@ -0,0 +1,1353 @@
...
+     * {@link #blemanagermonitorcharacteristicfordevice|bleManager.monitorCharacteristicForDevice()} with partially filled arguments.
+     *
+     * @param {function(error?: BleError, characteristic?: Characteristic)} listener callback which emits
+     * this {@link Characteristic} with modified value for each notification.
+     * @param {?TransactionId} transactionId optional `transactionId` which can be used in
+     * {@link #blemanagercanceltransaction|bleManager.cancelTransaction()} function.
+     * @returns {Subscription} Subscription on which `remove()` function can be called to unsubscribe.
+     */
+    monitor(
+      listener: (error?: BleError, characteristic?: Characteristic) => void,
```
Comment (Cierpliwy) [inline index.d.ts]: `listener: (error: BleError | null, characteristic: Characteristic | null) => void`

**Comment by shumih** [comment]
@Cierpliwy, guess that all errors are fixed

**Comment by Cierpliwy** [comment]
Thank you :) We will handle rest of the work (like CI/CD checks) and any other small issues related to definition file from that point.

**Pull request merged**